base/floatfuncs.jl

```diff
@@ -20,6 +20,7 @@ maxintfloat(::Type{Float64}) = 9007199254740992.
 maxintfloat(::Type{Float32}) = Float32(16777216.)
 maxintfloat(::Type{Float16}) = Float16(2048f0)
 maxintfloat(x::T) where {T<:AbstractFloat} = maxintfloat(T)
+maxintfloat(::Type{S}, ::Type{T}) where {S<:AbstractFloat, T<:Integer} = min(maxintfloat(S), S(typemax(T)))
```

Comment:
may be worth documenting and testing this new signature (including the corner cases that Tim pointed out), since it's exported if it's using the same name